File: examples/linux/anymeas_mode.cpp
/*
 * This example reads data from the Cirque trackpad in "anymeas mode" and prints the values.
 *
 * See documentation at https://cirquepinnacle.rtfd.io/
 */
#include <iostream>                        // cout, endl
#include <unistd.h>                        // sleep()
#include <CirquePinnacle/CirquePinnacle.h> // trackpad object

#ifdef PINNACLE_DRIVER_mraa
    #define DR_PIN 22 // GPIO25
#else
    #define DR_PIN 25
#endif
#define SS_PIN 0

#ifndef USE_I2C
PinnacleTouchSPI trackpad(DR_PIN, SS_PIN);
#else // If using I2C, then use the following line (not the line above)
PinnacleTouchI2C trackpad(DR_PIN);
#endif

typedef struct _MeasureVector
{
    unsigned long toggle;
    unsigned long polarity;
} measureVector;

measureVector vectorDeterminants[] = {
    // {toggle, polarity}
    {0x0000FFFF, 0x00000000}, // toggle all x-axis bits negatively (to 0)
    {0x0FFF0000, 0x00000000}, // toggle all y-axis bits negatively (to 0)
    {0x00FF00FF, 0x000000FF}, // toggle Y0-Y7 negatively (to 0) & X0-X7 positively (to 1)
    {0x00FF00FF, 0x00FF0000}  // toggle Y0-Y7 positively (to 1) & X0-X7 negatively (to 0)
};

const uint8_t variousVectors_size = sizeof(vectorDeterminants) / sizeof(measureVector);
int16_t compensations[variousVectors_size];

void compensate()
{
    signed long accumulatedValue;
    for (uint8_t i = 0; i < variousVectors_size; ++i) {
        uint8_t sweep = 0;
        accumulatedValue = 0;
        while (sweep < 5) // take 5 measurements and average them for a bit lower noise compensation value
        {
            int16_t value = trackpad.measureAdc(vectorDeterminants[i].toggle, vectorDeterminants[i].polarity);
            sweep++;
            accumulatedValue += value;
        }
        compensations[i] = accumulatedValue / 5;
        std::cout << "compensation " << (unsigned int)i << ": " << compensations[i] << std::endl;
    }
}

bool setup()
{
    if (!trackpad.begin()) {
        std::cout << "Cirque Pinnacle not responding!" << std::endl;
        return false;
    }
    std::cout << "CirquePinnacle/examples/linux/anymeas_mode" << std::endl;
    trackpad.setDataMode(PINNACLE_ANYMEAS);
    trackpad.anymeasModeConfig();
#ifndef USE_SW_DR // if using PINNACLE_SW_DR
    std::cout << "-- Using HW DataReady pin." << std::endl;
#endif
#ifndef USE_I2C
    std::cout << "-- Using SPI interface." << std::endl;
#else
    std::cout << "-- Using I2C interface." << std::endl;
#endif
    compensate();

    for (uint8_t i = 5; i; --i) {
        std::cout << "starting in " << (unsigned int)i << "second" << (i < 1 ? 's' : ' ') << '\r';
        sleep(1);
    }
    std::cout << std::endl;
    return true;
}

void loop()
{
    for (unsigned int i = 0; i < variousVectors_size; i++) {
        int16_t measurement = trackpad.measureAdc(
            vectorDeterminants[i].toggle,
            vectorDeterminants[i].polarity);
        measurement -= compensations[i];
        std::cout << "meas" << i << ": " << measurement << "\t";
    }
    std::cout << std::endl;
}

int main()
{
    if (!setup()) { // if trackpad.begin() failed
        return -1;  // fail fast
    }
    while (true) { // use ctrl+C to exit
        loop();
    }
    return 0; // we will never reach this
}

File: examples/cpython/anymeas_mode.py
"""
This example reads data from the Cirque trackpad in "anymeas mode" and
prints the values.

See documentation at https://cirquepinnacle.rtfd.io/
"""

import time
from typing import Union
from cirque_pinnacle import (
    PinnacleTouchSPI,
    PinnacleTouchI2C,  # noqa: imported but unused
    PINNACLE_ANYMEAS,
)

print("CirquePinnacle/examples/cpython/anymeas_mode\n")

# The digital input pin connected to the trackpad's DR (DataReady) pin
dr_pin = 25  # GPIO25

trackpad: Union[PinnacleTouchSPI, PinnacleTouchI2C]
if not input("Is the trackpad configured for I2C? [y/N] ").lower().startswith("y"):
    print("-- Using SPI interface.")
    ss_pin = 0  # uses /dev/spidev0.0 (CE0 or GPIO8)
    trackpad = PinnacleTouchSPI(dr_pin, ss_pin)
else:
    print("-- Using I2C interface")
    trackpad = PinnacleTouchI2C(dr_pin)

if not trackpad.begin():
    raise OSError("Cirque Pinnacle not responding!")
trackpad.data_mode = PINNACLE_ANYMEAS

# we'll use this list of tuples as args to PinnacleTouch.measure_adc()
vectors = [
    #  toggle  ,   polarity
    (0x0000FFFF, 0x00000000),  # toggle all x-axis bits negatively (to 0)
    (0x0FFF0000, 0x00000000),  # toggle all y-axis bits negatively (to 0)
    (
        0x00FF00FF,
        0x000000FF,
    ),  # toggle Y0-Y7 negatively (to 0) & X0-X7 positively (to 1)
    (
        0x00FF00FF,
        0x00FF0000,
    ),  # toggle Y0-Y7 positively (to 1) & X0-X7 negatively (to 0)
]

# a list of compensations to use with measured `vectors`
compensation = [0] * len(vectors)


def compensate():
    sweep = 5
    for i, (toggle, polarity) in enumerate(vectors):
        value = 0
        for _ in range(sweep):
            value += trackpad.measure_adc(toggle, polarity)
        compensation[i] = int(value / sweep)
        print("compensation {}: {}".format(i, compensation[i]))


def take_measurements(timeout=6):
    """Read ``len(vectors)`` number of measurements and print results for
    ``timeout`` number of seconds."""
    print("Taking measurements for", timeout, "seconds.")
    start = time.monotonic()
    while time.monotonic() - start < timeout:
        for i, (toggle, polarity) in enumerate(vectors):
            result = trackpad.measure_adc(toggle, polarity)
            print("meas{}: {}".format(i, result - compensation[i]), end="\t")
        print()


def set_role():
    """Set the role using stdin stream. Arguments for functions can be
    specified using a space delimiter (e.g. 'C 10' calls `compensate(10)`)
    """
    user_input = (
        input(
            "\n*** Enter 'C' to get compensations for measurements."
            "\n*** Enter 'M' to read and print measurements."
            "\n*** Enter 'Q' to quit example.\n"
        )
        or "?"
    ).split()
    if user_input[0].upper().startswith("C"):
        compensate(*[int(x) for x in user_input[1:2]])
        return True
    if user_input[0].upper().startswith("M"):
        take_measurements(*[int(x) for x in user_input[1:2]])
        return True
    if user_input[0].upper().startswith("Q"):
        return False
    print(user_input[0], "is an unrecognized input. Please try again.")
    return set_role()


if __name__ == "__main__":
    try:
        while set_role():
            pass  # continue example until 'Q' is entered
    except KeyboardInterrupt:
        print(" Keyboard Interrupt detected.")
else:
    print(
        "\nRun compensate() to set compensations for measurements.",
        "Run take_measurements() to read and print measurements.",
        sep="\n",
    )
